src/core/thread/network_data_tlvs.hpp

```diff
@@ -1000,148 +1020,62 @@ class ServiceTlv : public NetworkDataTlv
-                (length >= ((sizeof(*this) - sizeof(NetworkDataTlv)) + (IsThreadEnterprise() ? 0 : sizeof(uint32_t)) +
-                            sizeof(uint8_t))) &&
-                (length >= ((sizeof(*this) - sizeof(NetworkDataTlv)) + (IsThreadEnterprise() ? 0 : sizeof(uint32_t)) +
-                            sizeof(uint8_t) + GetServiceDataLength())));
-    }
-
-    /**
-     * This method gets Service Data length.
-     *
-     * @returns length of the Service Data field in bytes.
-     *
-     */
-    uint8_t GetServiceDataLength(void) const { return *GetServiceDataLengthLocation(); }
-
-    /**
-     * This method sets Service Data length.
-     *
-     * @param[in] aServiceDataLength desired length of the Service Data field in bytes.
-     *
-     */
-    void SetServiceDataLength(uint8_t aServiceDataLength) { *GetServiceDataLengthLocation() = aServiceDataLength; }
 
-    /**
-     * This method returns a pointer to the Service Data.
-     *
-     * @returns A pointer to the Service Data.
-     *
-     */
-    uint8_t *GetServiceData(void) { return GetServiceDataLengthLocation() + sizeof(uint8_t); }
-
-    /**
-     * This method returns a pointer to the Service Data.
-     *
-     * @returns A pointer to the Service Data.
-     *
-     */
-    const uint8_t *GetServiceData(void) const { return GetServiceDataLengthLocation() + sizeof(uint8_t); }
+        return (length >= sizeof(mFlagsServiceId)) &&
+               (length >= kMinLength + (IsThreadEnterprise() ? 0 : sizeof(mEnterpriseNumber))) &&
+               (length >= GetFieldsLength());
+    }
 
     /**
-     * This method sets Service Data to the given values.
-     *
-     * Caller must ensure that there is enough memory allocated.
+     * This method returns the Service ID. It is in range 0x00-0x0f.
      *
-     * @param aServiceData       pointer to the service data to use
-     * @param aServiceDataLength length of the provided service data in bytes
+     * @returns the Service ID.
      *
      */
-    void SetServiceData(const uint8_t *aServiceData, uint8_t aServiceDataLength)
-    {
-        SetServiceDataLength(aServiceDataLength);
-
-        memcpy(GetServiceData(), aServiceData, aServiceDataLength);
-    }
+    uint8_t GetServiceId(void) const { return (mFlagsServiceId & kServiceIdMask); }
 
     /**
      * This method returns Enterprise Number field.
      *
-     * @returns Enterprise Number
+     * @returns The Enterprise Number.
      *
      */
     uint32_t GetEnterpriseNumber(void) const
     {
-        if (IsThreadEnterprise())
-        {
-            return THREAD_ENTERPRISE_NUMBER;
-        }
-        else
-        {
-            // This memory access most likely will not be aligned to 4 bytes
-            return HostSwap32(*reinterpret_cast<const uint32_t *>(GetEnterpriseNumberLocation()));
-        }
+        return IsThreadEnterprise() ? static_cast<uint32_t>(kThreadEnterpriseNumber) : HostSwap32(mEnterpriseNumber);
```

Comment:
@dismirlian, I believe this addresses one of the items called out in #4796 .

Reply:
Yes. The new model should remove the need for casting the pointer and possible un-aligned access. I personally don't like the pointer casting approach for reading/writing any integer (other then `uint8_t`). The behavior is tool-chain dependent.

We have methods in `Encoding` like `ReadUint32`, `WriteUint32`, etc (for both little/big endian encodings) that do the same and should be safer and work with any pointer value (independent of its alignment).

Reply:
@jwhui yes it is one of those instances. I agree with both of you that `ReadUint32` (and similar) are the best way forward.  @abtink's refactoring gets rid of the possible unaligned access, so I'll hold my PR until this is merged. Thanks!